File: apps/android/app/src/main/java/com/knowtomigrate/app/network/KtmTransferClient.kt
package com.knowtomigrate.app.network

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.*
import java.net.InetSocketAddress
import java.net.Socket
import java.security.MessageDigest

class KtmTransferClient(private val context: Context) {

    private val _progress = MutableStateFlow(TransferProgressInfo())
    val progress: StateFlow<TransferProgressInfo> = _progress.asStateFlow()

    suspend fun sendUris(
        targetIp: String,
        targetPort: Int,
        targetDeviceName: String = targetIp,
        localDeviceId: String,
        localDeviceName: String,
        uris: List<Uri>,
        selectedTransport: String = "WIFI_LAN"
    ): Boolean = withContext(Dispatchers.IO) {
        if (uris.isEmpty()) return@withContext true

        val friendlyTransport = KtmTransportType.fromCode(selectedTransport).displayName
        val historyRepo = com.knowtomigrate.app.data.TransferHistoryRepository.getInstance(context)
        val prog = TransferProgressInfo(
            peerName = targetDeviceName,
            totalFiles = uris.size,
            transportType = friendlyTransport
        )

        try {
            Log.i("KtmTransferClient", "[TRANSFER_CREATED] Target=$targetIp:$targetPort, Files=${uris.size}")
            Socket().use { socket ->
                socket.tcpNoDelay = true
                socket.connect(InetSocketAddress(targetIp, targetPort), 10000)

                val inputStream = DataInputStream(BufferedInputStream(socket.getInputStream(), 64 * 1024))
                val outputStream = DataOutputStream(BufferedOutputStream(socket.getOutputStream(), 256 * 1024))

                // 1. Build Manifest from Uris
                val manifestItems = mutableListOf<KtmManifestItem>()
                var totalBytes = 0L

                uris.forEachIndexed { index, uri ->
                    val (name, size) = queryFileInfo(uri)
                    val sha = computeUriSha256(uri)
                    manifestItems.add(
                        KtmManifestItem(
                            fileIndex = index,
                            relativePath = name,
                            size = size,
                            sha256 = sha,
                            isFolder = false
                        )
                    )
                    totalBytes += size
                    Log.i("KtmTransferClient", "[MANIFEST_ITEM] #$index: '$name', size=$size bytes, SHA=$sha")
                }

                val manifest = KtmManifest(
                    senderDeviceId = localDeviceId,
                    senderDeviceName = localDeviceName,
                    totalBytes = totalBytes,
                    totalFiles = manifestItems.size,
                    files = manifestItems
                )

                prog.sessionId = manifest.sessionId
                prog.totalBytes = totalBytes
                _progress.value = prog.copy()

                val mainFileName = manifestItems.firstOrNull()?.relativePath ?: "Files"
                val displayTitle = if (manifestItems.size > 1) "$mainFileName (+${manifestItems.size - 1} more)" else mainFileName
                historyRepo.insert(
                    com.knowtomigrate.app.data.TransferRecord(
                        sessionId = manifest.sessionId,
                        fileName = displayTitle,
                        fileCount = manifestItems.size,
                        totalBytes = totalBytes,
                        bytesTransferred = 0L,
                        peerDeviceName = targetDeviceName,
                        direction = com.knowtomigrate.app.data.TransferDirection.SENT,
                        status = com.knowtomigrate.app.data.TransferRecordStatus.CONNECTING,
                        transport = selectedTransport
                    )
                )

                // 2. Handshake with 6-digit confirmation PIN
                val pin = KtmSecurityUtils.generate6DigitPin()
                val handshakeObj = JSONObject().apply {
                    put("type", "HANDSHAKE")
                    put("deviceId", localDeviceId)
                    put("deviceName", localDeviceName)
                    put("platform", "Android")
                    put("pin", pin)
                    put("selectedTransport", selectedTransport)
                }
                Log.i("KtmTransferClient", "[HANDSHAKE_SEND] PIN=$pin, Sender=$localDeviceName, Transport=$selectedTransport")
                writeLengthPrefixedString(outputStream, handshakeObj.toString())

                val ackJson = readLengthPrefixedString(inputStream)
                val ackObj = JSONObject(ackJson)
                if (!ackObj.optBoolean("accepted", false)) {
                    val reason = ackObj.optString("reason", "Rejected by recipient")
                    Log.w("KtmTransferClient", "[HANDSHAKE_REJECTED] Reason: $reason")
                    historyRepo.updateProgress(manifest.sessionId, com.knowtomigrate.app.data.TransferRecordStatus.CANCELLED, 0L, totalBytes, reason)
                    throw IOException("Transfer rejected by recipient: $reason")
                }
                Log.i("KtmTransferClient", "[HANDSHAKE_ACCEPTED] Recipient confirmed PIN")
                historyRepo.updateProgress(manifest.sessionId, com.knowtomigrate.app.data.TransferRecordStatus.TRANSFERRING, 0L, totalBytes)

                // 3. Send Manifest
                writeLengthPrefixedString(outputStream, manifest.toJson())

                val manifestAckJson = readLengthPrefixedString(inputStream)
                val manifestAckObj = JSONObject(manifestAckJson)
                if (!manifestAckObj.optBoolean("accepted", false)) {
                    val reason = manifestAckObj.optString("reason", "Manifest rejected")
                    Log.w("KtmTransferClient", "[MANIFEST_REJECTED] Reason: $reason")
                    throw IOException("Manifest rejected by recipient: $reason")
                }

                val existingOffsets = manifestAckObj.optJSONObject("existingOffsets") ?: JSONObject()
                Log.i("KtmTransferClient", "[MANIFEST_ACCEPTED] SessionId=${manifest.sessionId}, Offsets=$existingOffsets")

                // 4. Stream Chunks with Resume Support
                val chunkBuffer = ByteArray(KtmConstants.DEFAULT_CHUNK_SIZE)
                var lastTimerTime = System.currentTimeMillis()
                var bytesSinceTimer = 0L

                manifest.files.forEachIndexed { index, item ->
                    val uri = uris[index]
                    val resumeOffset = existingOffsets.optLong(item.fileIndex.toString(), 0L)
                    var fileBytesSent = resumeOffset

                    prog.currentFileName = item.relativePath
                    prog.currentFileIndex = index + 1
                    _progress.value = prog.copy()

                    if (item.size > 0L) {
                        val stream = context.contentResolver.openInputStream(uri)
                            ?: throw IOException("Cannot open input stream for ${item.relativePath} (URI: $uri)")

                        stream.use { inStream ->
                            Log.i("KtmTransferClient", "[FILE_OPEN_SUCCESS] '${item.relativePath}', resumeOffset=$resumeOffset, targetSize=${item.size}")
                            if (resumeOffset > 0L) {
                                var skipped = 0L
                                while (skipped < resumeOffset) {
                                    val s = inStream.skip(resumeOffset - skipped)
                                    if (s <= 0L) break
                                    skipped += s
                                }
                                prog.bytesTransferred += resumeOffset
                            }

                            while (fileBytesSent < item.size) {
                                val toRead = Math.min(chunkBuffer.size.toLong(), item.size - fileBytesSent).toInt()
                                val read = inStream.read(chunkBuffer, 0, toRead)
                                if (read <= 0) break

                                // Write 20-byte frame header (Big-Endian)
                                outputStream.writeInt(KtmConstants.CHUNK_MAGIC)
                                outputStream.writeInt(item.fileIndex)
                                outputStream.writeLong(fileBytesSent)
                                outputStream.writeInt(read)
                                outputStream.write(chunkBuffer, 0, read)
                                outputStream.flush()

                                fileBytesSent += read
                                prog.bytesTransferred += read
                                bytesSinceTimer += read

                                val now = System.currentTimeMillis()
                                if (now - lastTimerTime >= 500) {
                                    val secs = (now - lastTimerTime) / 1000.0
                                    prog.speedMBps = (bytesSinceTimer / (1024.0 * 1024.0)) / secs
                                    lastTimerTime = now
                                    bytesSinceTimer = 0L
                                    _progress.value = prog.copy()
                                }
                            }
                        }
                        Log.i("KtmTransferClient", "[BYTES_SENT] '${item.relativePath}': $fileBytesSent/${item.size} bytes sent")
                    } else {
                        Log.i("KtmTransferClient", "[ZERO_BYTE_FILE] '${item.relativePath}' is 0 bytes, skipping chunk streaming")
                    }

                    // Read FILE_COMPLETE ack from receiver
                    val compJson = readLengthPrefixedString(inputStream)
                    val compObj = JSONObject(compJson)
                    val status = compObj.optString("status", "")
                    if (status != "OK") {
                        val err = "Recipient failed verification for ${item.relativePath} (status: $status)"
                        Log.e("KtmTransferClient", "[FILE_VERIFY_FAILED] $err")
                        throw IOException(err)
                    }
                    Log.i("KtmTransferClient", "[FILE_VERIFIED] '${item.relativePath}' confirmed OK by recipient")
                }

                // 5. Transfer Complete
                val doneJson = readLengthPrefixedString(inputStream)
                val doneObj = JSONObject(doneJson)

                prog.isCompleted = doneObj.optBoolean("success", false)
                _progress.value = prog.copy()
                Log.i("KtmTransferClient", "[TRANSFER_COMPLETED] Success=${prog.isCompleted}")
                if (prog.isCompleted) {
                    historyRepo.updateProgress(manifest.sessionId, com.knowtomigrate.app.data.TransferRecordStatus.COMPLETED, totalBytes, totalBytes)
                } else {
                    historyRepo.updateProgress(manifest.sessionId, com.knowtomigrate.app.data.TransferRecordStatus.FAILED, prog.bytesTransferred, totalBytes, "Transfer failed at destination")
                }
                return@withContext prog.isCompleted
            }
        } catch (e: Exception) {
            Log.e("KtmTransferClient", "[TRANSFER_ERROR] ${e.message}", e)
            prog.errorMessage = e.message ?: "Transfer error"
            _progress.value = prog.copy()
            if (prog.sessionId.isNotBlank()) {
                historyRepo.updateProgress(prog.sessionId, com.knowtomigrate.app.data.TransferRecordStatus.FAILED, prog.bytesTransferred, prog.totalBytes, e.message)
            }
            return@withContext false
        }
    }

    private fun queryFileInfo(uri: Uri): Pair<String, Long> {
        var name = "file_${System.currentTimeMillis()}"
        var size = -1L

        // Primary: Query Android ContentResolver OpenableColumns
        try {
            context.contentResolver.query(uri, null, null, null, null)?.use { cursor ->
                val nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                val sizeIndex = cursor.getColumnIndex(OpenableColumns.SIZE)
                if (cursor.moveToFirst()) {
                    if (nameIndex != -1 && !cursor.isNull(nameIndex)) {
                        val n = cursor.getString(nameIndex)
                        if (!n.isNullOrBlank()) name = n
                    }
                    if (sizeIndex != -1 && !cursor.isNull(sizeIndex)) {
                        val s = cursor.getLong(sizeIndex)
                        if (s >= 0) size = s
                    }
                }
            }
        } catch (e: Exception) {
            Log.w("KtmTransferClient", "Cursor query failed for $uri", e)
        }

        // Fallback 1: AssetFileDescriptor length
        if (size <= 0) {
            try {
                context.contentResolver.openAssetFileDescriptor(uri, "r")?.use { afd ->
                    val afdLen = afd.length
                    if (afdLen > 0) size = afdLen
                }
            } catch (_: Exception) {}
        }

        // Fallback 2: ParcelFileDescriptor statSize
        if (size <= 0) {
            try {
                context.contentResolver.openFileDescriptor(uri, "r")?.use { pfd ->
                    val pfdSize = pfd.statSize
                    if (pfdSize > 0) size = pfdSize
                }
            } catch (_: Exception) {}
        }

        // Fallback 3: Measure stream length directly
        if (size <= 0) {
            try {
                context.contentResolver.openInputStream(uri)?.use { stream ->
                    var count = 0L
                    val buf = ByteArray(65536)
                    var r: Int
                    while (stream.read(buf).also { r = it } != -1) {
                        count += r
                    }
                    size = count
                }
            } catch (e: Exception) {
                Log.w("KtmTransferClient", "Stream byte measurement failed for $uri", e)
            }
        }

        // Name fallback from URI path if generic
        if (name.startsWith("file_")) {
            val lastSegment = uri.lastPathSegment
            if (!lastSegment.isNullOrBlank()) {
                val clean = lastSegment.substringAfterLast('/').substringAfterLast(':')
                if (clean.isNotBlank()) name = clean
            }
        }

        if (size < 0) size = 0L
        return Pair(name, size)
    }

    private fun computeUriSha256(uri: Uri): String {
        return try {
            val digest = MessageDigest.getInstance("SHA-256")
            context.contentResolver.openInputStream(uri)?.use { stream ->
                val buf = ByteArray(65536)
                var r: Int
                while (stream.read(buf).also { r = it } != -1) {
                    digest.update(buf, 0, r)
                }
            }
            digest.digest().joinToString("") { "%02X".format(it) }
        } catch (e: Exception) {
            Log.w("KtmTransferClient", "Failed to compute SHA-256 for $uri", e)
            ""
        }
    }

    private fun readLengthPrefixedString(dis: DataInputStream): String {
        val len = dis.readInt()
        if (len <= 0 || len > 10 * 1024 * 1024) throw IOException("Invalid string length: $len")
        val bytes = ByteArray(len)
        dis.readFully(bytes)
        return String(bytes, Charsets.UTF_8)
    }

    private fun writeLengthPrefixedString(dos: DataOutputStream, str: String) {
        val bytes = str.toByteArray(Charsets.UTF_8)
        dos.writeInt(bytes.size)
        dos.write(bytes)
        dos.flush()
    }
}

File: apps/windows-wpf/Services/KtmTransferClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace KnowToMigrate.Services
{
    public class KtmTransferClient
    {
        public event Action<TransferProgressInfo>? OnProgress;

        public async Task<bool> SendFilesAsync(
            string targetIp,
            int targetPort,
            string localDeviceId,
            string localDeviceName,
            IReadOnlyList<string> filePaths,
            string selectedTransport = KtmTransportCodes.WifiLan,
            CancellationToken token = default)
        {
            if (filePaths.Count == 0) return true;

            using var client = new TcpClient();
            client.ReceiveBufferSize = 64 * 1024;
            client.SendBufferSize = 256 * 1024;
            client.NoDelay = true;

            long startMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var progress = new TransferProgressInfo
            {
                PeerName = targetIp,
                TotalFiles = 0,
                TotalBytes = 0,
                TransportType = KtmTransportCodes.GetDisplayName(selectedTransport),
                Direction = TransferDirection.Sending,
                Status = TransferStatus.Connecting,
                StartTimeMs = startMs
            };

            try
            {
                await client.ConnectAsync(targetIp, targetPort, token);
                using var stream = client.GetStream();

                // 1. Build Manifest
                var manifest = new KtmManifest
                {
                    SessionId = Guid.NewGuid().ToString("N"),
                    SenderDeviceId = localDeviceId,
                    SenderDeviceName = localDeviceName
                };

                int fileIdx = 0;
                long totalBytes = 0;

                foreach (var path in filePaths)
                {
                    if (File.Exists(path))
                    {
                        var fi = new FileInfo(path);
                        string sha = KtmSecurityUtils.ComputeFileSha256(path);
                        manifest.Files.Add(new KtmManifestItem
                        {
                            FileIndex = fileIdx++,
                            RelativePath = Path.GetFileName(path),
                            Size = fi.Length,
                            Sha256 = sha,
                            IsFolder = false
                        });
                        totalBytes += fi.Length;
                    }
                    else if (Directory.Exists(path))
                    {
                        var dirInfo = new DirectoryInfo(path);
                        string baseDirName = dirInfo.Name;
                        foreach (var subFile in Directory.EnumerateFiles(path, "*", SearchOption.AllDirectories))
                        {
                            var fi = new FileInfo(subFile);
                            string rel = Path.Combine(baseDirName, Path.GetRelativePath(path, subFile));
                            string sha = KtmSecurityUtils.ComputeFileSha256(subFile);
                            manifest.Files.Add(new KtmManifestItem
                            {
                                FileIndex = fileIdx++,
                                RelativePath = rel,
                                Size = fi.Length,
                                Sha256 = sha,
                                IsFolder = false
                            });
                            totalBytes += fi.Length;
                        }
                    }
                }

                manifest.TotalFiles = manifest.Files.Count;
                manifest.TotalBytes = totalBytes;
                progress.SessionId = manifest.SessionId;
                progress.TotalFiles = manifest.TotalFiles;
                progress.TotalBytes = totalBytes;
                progress.Status = TransferStatus.Connecting;
                OnProgress?.Invoke(progress);

                // 2. Handshake with 6-digit confirmation PIN
                string pin = KtmSecurityUtils.Generate6DigitPin();
                var handshake = new KtmHandshake
                {
                    DeviceId = localDeviceId,
                    DeviceName = localDeviceName,
                    Platform = "Windows",
                    Pin = pin,
                    SelectedTransport = selectedTransport
                };

                await SendLengthPrefixedJsonAsync(stream, handshake, token);
                string ackJson = await ReadLengthPrefixedJsonAsync(stream, token);
                var handshakeAck = JsonSerializer.Deserialize<KtmHandshakeAck>(ackJson);
                if (handshakeAck == null || !handshakeAck.Accepted)
                {
                    string reason = handshakeAck?.Reason ?? "Unknown";
                    progress.Status = TransferStatus.Failed;
                    progress.ErrorMessage = reason;
                    OnProgress?.Invoke(progress);
                    throw new InvalidOperationException($"Transfer rejected by recipient: {reason}");
                }

                progress.Status = TransferStatus.Transferring;
                OnProgress?.Invoke(progress);

                // 3. Send Manifest
                await SendLengthPrefixedJsonAsync(stream, manifest, token);
                string manifestAckJson = await ReadLengthPrefixedJsonAsync(stream, token);
                var manifestAck = JsonSerializer.Deserialize<KtmManifestAck>(manifestAckJson);
                if (manifestAck == null || !manifestAck.Accepted)
                {
                    string reason = manifestAck?.Reason ?? "Unknown";
                    progress.Status = TransferStatus.Failed;
                    progress.ErrorMessage = reason;
                    OnProgress?.Invoke(progress);
                    throw new InvalidOperationException($"Manifest rejected by recipient: {reason}");
                }

                // 4. Stream Chunks with Resume Support
                var existingOffsets = manifestAck.ExistingOffsets ?? new Dictionary<int, long>();
                byte[] chunkBuffer = new byte[KtmConstants.DefaultChunkSize];
                var speedTimer = System.Diagnostics.Stopwatch.StartNew();
                long bytesSinceTimer = 0;

                // Map file index back to local source path
                var indexToSourcePath = new Dictionary<int, string>();
                int mapIdx = 0;
                foreach (var path in filePaths)
                {
                    if (File.Exists(path))
                    {
                        indexToSourcePath[mapIdx++] = path;
                    }
                    else if (Directory.Exists(path))
                    {
                        foreach (var subFile in Directory.EnumerateFiles(path, "*", SearchOption.AllDirectories))
                        {
                            indexToSourcePath[mapIdx++] = subFile;
                        }
                    }
                }

                foreach (var fileItem in manifest.Files)
                {
                    if (token.IsCancellationRequested) break;

                    string localSource = indexToSourcePath[fileItem.FileIndex];
                    long resumeOffset = existingOffsets.TryGetValue(fileItem.FileIndex, out long ro) ? ro : 0;

                    progress.CurrentFileName = fileItem.RelativePath;
                    progress.CurrentFileIndex = fileItem.FileIndex + 1;
                    progress.Status = TransferStatus.Transferring;
                    OnProgress?.Invoke(progress);

                    using (var fs = new FileStream(localSource, FileMode.Open, FileAccess.Read, FileShare.Read, 131072, useAsync: true))
                    {
                        if (resumeOffset > 0 && resumeOffset < fileItem.Size)
                        {
                            fs.Seek(resumeOffset, SeekOrigin.Begin);
                            progress.BytesTransferred += resumeOffset;
                        }

                        long fileBytesSent = resumeOffset;
                        while (fileBytesSent < fileItem.Size)
                        {
                            if (token.IsCancellationRequested) break;

                            int toRead = (int)Math.Min((long)chunkBuffer.Length, fileItem.Size - fileBytesSent);
                            int read = await fs.ReadAsync(chunkBuffer, 0, toRead, token);
                            if (read <= 0) break;

                            // Build frame header (20 bytes Big-Endian)
                            byte[] frameHeader = new byte[20];
                            System.Buffers.Binary.BinaryPrimitives.WriteUInt32BigEndian(frameHeader.AsSpan(0, 4), KtmConstants.ChunkMagic);
                            System.Buffers.Binary.BinaryPrimitives.WriteInt32BigEndian(frameHeader.AsSpan(4, 4), fileItem.FileIndex);
                            System.Buffers.Binary.BinaryPrimitives.WriteInt64BigEndian(frameHeader.AsSpan(8, 8), fileBytesSent);
                            System.Buffers.Binary.BinaryPrimitives.WriteInt32BigEndian(frameHeader.AsSpan(16, 4), read);

                            await stream.WriteAsync(frameHeader, 0, 20, token);
                            await stream.WriteAsync(chunkBuffer, 0, read, token);
                            await stream.FlushAsync(token);

                            fileBytesSent += read;
                            progress.BytesTransferred += read;
                            bytesSinceTimer += read;
                            progress.ElapsedTimeMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - startMs;

                            if (speedTimer.ElapsedMilliseconds >= 300 || bytesSinceTimer >= 512 * 1024)
                            {
                                double secs = Math.Max(0.05, speedTimer.ElapsedMilliseconds / 1000.0);
                                progress.SpeedMBps = (bytesSinceTimer / (1024.0 * 1024.0)) / secs;
                                speedTimer.Restart();
                                bytesSinceTimer = 0;
                                OnProgress?.Invoke(progress);
                            }
                        }
                    }

                    // Read FILE_COMPLETE ack
                    progress.Status = TransferStatus.Verifying;
                    OnProgress?.Invoke(progress);

                    string fileCompleteJson = await ReadLengthPrefixedJsonAsync(stream, token);
                    var fileComp = JsonSerializer.Deserialize<KtmFileComplete>(fileCompleteJson);
                    if (fileComp == null || fileComp.Status != "OK")
                    {
                        progress.Status = TransferStatus.Failed;
                        progress.ErrorMessage = $"Checksum mismatch on recipient for {fileItem.RelativePath}";
                        OnProgress?.Invoke(progress);
                        throw new CryptographicException($"Checksum mismatch on recipient for {fileItem.RelativePath}");
                    }
                }

                // 5. Transfer Complete
                string transferCompleteJson = await ReadLengthPrefixedJsonAsync(stream, token);
                var transferComp = JsonSerializer.Deserialize<KtmTransferComplete>(transferCompleteJson);

                progress.IsCompleted = transferComp != null && transferComp.Success;
                progress.Status = progress.IsCompleted ? TransferStatus.Completed : TransferStatus.Failed;
                progress.ElapsedTimeMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - startMs;
                OnProgress?.Invoke(progress);
                return progress.IsCompleted;
            }
            catch (Exception ex)
            {
                progress.Status = TransferStatus.Failed;
                progress.ErrorMessage = ex.Message;
                OnProgress?.Invoke(progress);
                return false;
            }
        }

        private static async Task<string> ReadLengthPrefixedJsonAsync(NetworkStream stream, CancellationToken token)
        {
            byte[] lenBytes = new byte[4];
            await ReadExactBytesAsync(stream, lenBytes, 0, 4, token);
            int len = (lenBytes[0] << 24) | (lenBytes[1] << 16) | (lenBytes[2] << 8) | lenBytes[3];

            if (len <= 0 || len > 10 * 1024 * 1024)
                throw new InvalidDataException($"Invalid packet length: {len}");

            byte[] jsonBytes = new byte[len];
            await ReadExactBytesAsync(stream, jsonBytes, 0, len, token);
            return Encoding.UTF8.GetString(jsonBytes);
        }

        private static async Task SendLengthPrefixedJsonAsync<T>(NetworkStream stream, T obj, CancellationToken token)
        {
            string json = JsonSerializer.Serialize(obj);
            byte[] jsonBytes = Encoding.UTF8.GetBytes(json);
            int len = jsonBytes.Length;

            byte[] header = new byte[4];
            header[0] = (byte)((len >> 24) & 0xFF);
            header[1] = (byte)((len >> 16) & 0xFF);
            header[2] = (byte)((len >> 8) & 0xFF);
            header[3] = (byte)(len & 0xFF);

            await stream.WriteAsync(header, 0, 4, token);
            await stream.WriteAsync(jsonBytes, 0, len, token);
            await stream.FlushAsync(token);
        }

        private static async Task ReadExactBytesAsync(Stream stream, byte[] buffer, int offset, int count, CancellationToken token)
        {
            int totalRead = 0;
            while (totalRead < count)
            {
                int read = await stream.ReadAsync(buffer, offset + totalRead, count - totalRead, token);
                if (read == 0)
                    throw new EndOfStreamException("Remote host closed connection prematurely");
                totalRead += read;
            }
        }
    }
}
